activerecord/lib/active_record/insert_all.rb

```diff
@@ -61,6 +60,8 @@ def find_unique_index_for(unique_by)
 
         if index = unique_indexes.find { |i| match.include?(i.name) || i.columns == match }
           index
+        elsif match == primary_keys
```

Comment:
I don't understand the purpose of this? Also: what happens with different sort orders?

Reply:
The current implementation also errors if the columns passed in aren't in the same order as the index. I think this requires some further work.

Reply:
If the primary key is passed in - we don't have a Index object (responding to columns and where) that we could use to generate the sql clause in `conflict_target`. 

Having `unique_by` be nil and `@on_duplicate == :update` was how it said that we should use the primary key. 

Without this condition we get an error if we pass in the primary key to the `unique_by`